libs/wingc/src/type_check.rs

```diff
@@ -2512,32 +2503,23 @@ impl<'a> TypeChecker<'a> {
 			self.validate_structural_type(&arg_list_types.named_args, &last_arg, call_span);
 		}
 
-		// Count number of optional parameters from the end of the function's params
-		// Allow arg_list to be missing up to that number of nil values to try and make the number of arguments match
-		let num_optionals = func_sig
-			.parameters
-			.iter()
-			.rev()
-			.take_while(|arg| arg.typeref.is_option())
-			.count();
+		// Check if there is a variadic parameter, get its index
+		let variadic_index = func_sig.parameters.iter().position(|o| o.variadic);
+		let non_variadic_len = cmp::min(
+			arg_list.pos_args.len(),
+			variadic_index.unwrap_or(arg_list.pos_args.len()),
+		);
 
 		// Verify arity
-
-		// check if there is a variadic parameter, get its index
-		let variadic_index = func_sig.parameters.iter().position(|o| o.variadic);
-		let (index_last_item, arg_count) = if let Some(variadic_index) = variadic_index {
-			(
-				variadic_index,
-				(variadic_index + 1) + (if arg_list.named_args.is_empty() { 0 } else { 1 }),
-			)
-		} else {
-			(
-				arg_list_types.pos_args.len(),
-				(arg_list_types.pos_args.len()) + (if arg_list.named_args.is_empty() { 0 } else { 1 }),
-			)
-		};
-		let min_args = func_sig.parameters.len() - num_optionals;
+		let min_args = func_sig.min_parameters();
 		let max_args = func_sig.parameters.len();
+		let arg_count = (non_variadic_len)
+			+ (if variadic_index.is_some() && arg_list.pos_args.len() > non_variadic_len {
+				1
+			} else {
+				0
+			}) + (if arg_list.named_args.is_empty() { 0 } else { 1 });
+
 		if arg_count < min_args || arg_count > max_args {
 			let err_text = if min_args == max_args {
 				format!("Expected {} arguments but got {}", min_args, arg_count)
```

Comment:
I know you didn't change this part, but the case of `Expected 1 arguments` is very noticeable now :)

```suggestion
				format!("Expected {} argument(s) but got {}", min_args, arg_count)
```

Reply:
true, nice catch!